### PR TITLE
Created VMHookFactory

### LINKED ISSUE #1070: Refactor VMHook
VM hooks are currently not thread safe due to sharing one `VMHook` instance across all `VM` instances. For example, it turns into concurrent modifications because of `PendingState` processing. A straightforward way of solving this issue is to instantiate own `VMHook` instance for each `VM`. One of the way to achieve this is to introduce `VMHookFactory`

### PR DESCRIPTION
**Pull request by RichardH92**
Closes #1070

### CODE REVIEW

File: ethereumj-core/src/main/java/org/ethereum/vm/VM.java [hunk 1]
```diff
@@ -1328,8 +1333,8 @@ public void play(Program program) {
         }
     }
 
-    public static void setVmHook(VMHook vmHook) {
-        VM.vmHook = vmHook;
+    public static void setVmHookFactory(VMHookFactory vmHookFactory) {
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/vm/VM.java]: Let's keep old `VMHook` behavior for backward compatibility and mark it as deprecated with pointing to new alternative with detailed description of why client should use new feature

File: ethereumj-core/src/main/java/org/ethereum/vm/VMHookFactory.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package org.ethereum.vm;
+
+/**
+ * Created by Richard Hart on 07.07.2018.
+ */
+public interface VMHookFactory {
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/vm/VMHookFactory.java]: Please, provide a description of what this class is needed for

File: ethereumj-core/src/main/java/org/ethereum/vm/VMHookFactory.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package org.ethereum.vm;
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/vm/VMHookFactory.java]: Lack of license header

File: ethereumj-core/src/main/java/org/ethereum/vm/VMHookFactory.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+
+/**
+ * Created by Richard Hart on 07.07.2018.
+ */
+public interface VMHookFactory {
+    VMHook build();
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/vm/VMHookFactory.java]: Let's use `create` naming for this method.

**Comment by RichardH92** [comment]
@mkalinin Thanks for the feedback! I think this is ready for re-review, if you don't mind.

File: ethereumj-core/src/main/java/org/ethereum/vm/VM.java [hunk 5]
```diff
@@ -367,6 +374,8 @@ else if (oldValue != null && newValue.isZero()) {
 
             if (vmHook != null) {
                 vmHook.step(program, op);
+            } else if (globalVmHook != null) {
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/vm/VM.java]: `globalVmHook` must be triggered even if `vmHookFactory` is defined

File: ethereumj-core/src/main/java/org/ethereum/vm/VM.java [hunk 6]
```diff
@@ -20,6 +20,7 @@
 import org.ethereum.config.BlockchainConfig;
 import org.ethereum.config.SystemProperties;
 import org.ethereum.db.ContractDetails;
+import org.ethereum.sync.FastSyncManager;
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/vm/VM.java]: Is that dependency needed here?

File: ethereumj-core/src/main/java/org/ethereum/config/SystemProperties.java [hunk 7]
```diff
@@ -37,6 +37,7 @@
 import org.ethereum.util.Utils;
 import org.ethereum.validator.BlockCustomHashRule;
 import org.ethereum.validator.BlockHeaderValidator;
+import org.ethereum.vm.VMHook;
```
Comment (mkalinin) [inline ethereumj-core/src/main/java/org/ethereum/config/SystemProperties.java]: Looks like redundant dependency

**Comment by RichardH92** [comment]
Hey @mkalinin I believe I have addressed your 2nd round of comments. Please lmk if I need to change anything else

**Comment by zilm13** [comment]
Looks good to me, but I'd also like to see vmHooks calls to be bounded with try/catch so vmHook couldn't fail VM execution or is it excess protection, @mkalinin

**Comment by RichardH92** [comment]
@zilm13 That's a good idea. Just to be clear, you want me to catch all exceptions, right? Since this will be implemented by a third party? Should I just log the stacktrace?

**Comment by zilm13** [comment]
@RichardH92 yeah, just try-catch-log.error("... {}", exception), separate contours for `vmHook` and `globalVmHook` as we are going to remove last with time and to exclude interfere.

**Comment by RichardH92** [comment]
Hey @zilm13 how does this look to you? Any suggestions on the error messages?

**Comment by zilm13** [comment]
@RichardH92 How about adding something like this?
1) Don't like duplicates
2) You are not logging exceptions
I think it could be fixed with something like this:
```
    private void callVmHookAction(Program program, BiConsumer<VMHook, Program> action) {
        if (vmHook != null) {
            try {
                action.accept(vmHook, program);
            } catch (Exception e) {
                logger.error("Error calling VMHook action: {}", e);
            }
        }

        if (globalVmHook != null) {
            try {
                action.accept(globalVmHook, program);
            } catch (Exception e) {
                logger.error("Error calling global VMHook action: {}", e);
            }
        }
    }
```

and then call it:
```
callVmHookAction(program, (hook, prg) -> hook.step(prg, op));
callVmHookAction(program, VMHook::startPlay);
callVmHookAction(program, VMHook::stopPlay);
```

also, please, add exception logging to `.create` with smth like this:
```
logger.error("Error creating VMHook: {}", e);
```

**Comment by RichardH92** [comment]
Hey @zilm13 , thanks for the feedback. I think this is ready for re-review when you have time

File: ethereumj-core/src/main/java/org/ethereum/vm/VM.java [hunk 8]
```diff
@@ -375,7 +376,7 @@ else if (oldValue != null && newValue.isZero()) {
             if (program.getNumber().intValue() == dumpBlock)
                 this.dumpLine(op, gasBefore, gasCost + callGas, memWords, program);
 
-            vmHookStepWrapper(program, op);
+            callVmHookAction(program, (hook, prg) -> vmHook.step(prg, op));
```
Comment (zilm13) [inline ethereumj-core/src/main/java/org/ethereum/vm/VM.java]: `(hook, prg) -> vmHook.step(prg, op)` should be changed to `(hook, prg) -> hook.step(prg, op)` or you will call it twice in wrapper

**Comment by RichardH92** [comment]
@zilm13 Nice catch! Should be fixed now

**Comment by zilm13** [comment]
@RichardH92 Cool, thx!

**Pull request merged**